### ISSUE TITLE
Get [Tensor (undefined) ] when calling tensor.grad() in c++ front-end

### ISSUE THREAD

**Issue by JunhongXu** [pytorch/pytorch#19173]
## Issue

As shown in the following example, the gradient is undefined after we call tensor.view().

### Case One: without calling  view() on input tensors.

```
torch::Tensor a = torch::tensor({2, 1}, torch::TensorOptions().
		requires_grad(true).dtype(torch::kFloat32));
torch::Tensor b = torch::tensor({3, 3}, torch::TensorOptions().
		requires_grad(true).dtype(torch::kFloat32));
auto c = a * b;
c.backward(); 
std::cout<<"Gradient of a"<<std::endl; 
std::cout<<a.grad()<<std::endl; // gives 3, 3 as expected
std::cout<<"Gradient of b"<<std::endl;
std::cout<<b.grad()<<std::endl; // gives 2, 1 as expected
```

### Case Two: calling  view() on input tensors.

```
torch::Tensor a = torch::tensor({2, 1}, torch::TensorOptions().
		requires_grad(true).dtype(torch::kFloat32)).view({2, 1});
torch::Tensor b = torch::tensor({3, 3}, torch::TensorOptions().
		requires_grad(true).dtype(torch::kFloat32)).view({2, 1});
auto c = a * b;
c.backward(); 
std::cout<<"Gradient of a"<<std::endl; 
std::cout<<a.grad()<<std::endl; // gives undefined tensor
std::cout<<"Gradient of b"<<std::endl;
std::cout<<b.grad()<<std::endl; // gives undefined tensor
```

**Comment by SsnL** [comment]
This is a leaf, which can receive grad:
```
torch::Tensor b = torch::tensor({3, 3}, torch::TensorOptions().
		requires_grad(true);
```

What you had, is an intermediate result:
```
torch::Tensor b = torch::tensor({3, 3}, torch::TensorOptions().
		requires_grad(true).dtype(torch::kFloat32)).view({2, 1});
```
